src/combinatorics/quiver.lean

```diff
@@ -3,130 +3,139 @@ Copyright (c) 2021 David Wärn. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: David Wärn
 -/
-import tactic
+import data.equiv.basic
+import order.well_founded
+import data.nat.basic
+import data.opposite
+
 /-!
 # Quivers
 
-This module defines quivers. A quiver `G` on a type `V` of vertices assigns to every
-pair `a b : V` of vertices a type `G.arrow a b` of arrows from `a` to `b`. This
+This module defines quivers. A quiver on a type `V` of vertices assigns to every
+pair `a b : V` of vertices a type `a ⟶ b` of arrows from `a` to `b`. This
 is a very permissive notion of directed graph.
+
+## Implementation notes
+
+It would be interesting to try to replace `has_hom` with `quiver` in the definition of a category.
+This would be convenient for defining path categories.
+
 -/
 
+open opposite
+
 -- We use the same universe order as in category theory.
 -- See note [category_theory universes]
 universes v u
 
 /-- A quiver `G` on a type `V` of vertices assigns to every pair `a b : V` of vertices
-    a type `G.arrow a b` of arrows from `a` to `b`. -/
-structure quiver (V : Type u) :=
-(arrow : V → V → Sort v)
+    a type `a ⟶ b` of arrows from `a` to `b`. -/
+class quiver (V : Type u) :=
+(hom : V → V → Type v)
+
+infixr ` ⟶ `:10 := quiver.hom -- type as \h
 
 /-- A wide subquiver `H` of `G` picks out a set `H a b` of arrows from `a` to `b`
     for every pair of vertices `a b`. -/
-def wide_subquiver {V} (G : quiver V) :=
-Π a b : V, set (G.arrow a b)
+def wide_subquiver (V) [quiver V] :=
+Π a b : V, set (a ⟶ b)
+
+/-- A type synonym for `V`, when thought of as a quiver having only the arrows from
+some `wide_subquiver`. -/
+@[nolint unused_arguments has_inhabited_instance]
+def wide_subquiver_synonym (V) [quiver V] (H : wide_subquiver V) : Type u := V
```

Comment:
I wonder if we can come up with a better name for this? Maybe `wide_subquiver.to_Type` or something similar? Writing `H.to_Type` feels nicer than `wide_subquiver_synonym H`. OTOH, I understand that the intention is to use the coercion to type below...

Reply:
Good idea.